DM-28844: Switch to using click and add indexing command

timj opened the pull request:
* Add astrometadata click command
* Add translate, dump, write-sidecar, write-index, write-metadata-index subcommands

timj committed:
Factor out common code from translate header implementation

timj committed:
Convert translate_header to click

astrometadata translate
astrometadata dump

Uses shared implementation but now two distinct commands.

translate_header.py still exists.

timj committed:
Add write-sidecar subcommand

timj committed:
Add write-index command

timj committed:
Add diff mode for header merging

timj committed:
Improve some messages

timj committed:
Add write-metadata-index subcommand

timj committed:
Force PropertyList to dict if we are having to store diffs in it

timj committed:
Change index output format to have __COMMON__ and sorted files

timj committed:
Reorganize indexing code to allow tests to be written

Also adds a small test fits file.

timj committed:
Do not fixup header when writing metadata content

Doing so can confuse subsequent reads of the file and also
means that the content would have to be recalculated if the
fix logic changes.

timj committed:
Add write-metadata-sidecar and allow both sidecar types to be read

timj committed:
Add outpath override for index writing

This writes a summary of all files in one index file rather than
one per directory.

timj committed:
Add METADATA_TRANSLATORS env var for command line tools

This will automatically import additional packages.

timj committed:
Fix sphinx warning in file

timj committed:
Modernize the sphinx config for new documenteer

timj committed:
Document astrometadata command

timj committed:
Fix spelling error

timj committed:
Add astrometadata command line entry point

timj committed:
Make astropy fits reader match afw behavior on failure

timj committed:
Protect against bad translator package imports from command line

timj committed:
Allow index reading to always return dict

Sometimes you really do want a dict indexed by file name in
both cases.

timj committed:
Add note on relative paths in index files

fred3m commented on .gitattributes:
```diff
@@ -0,0 +1,2 @@
+*.fits binary
+*.npy binary
```
I never used `.gitattributes` before, so I obviously didn't know about the `binary` attribute. Nice to know!

fred3m commented on doc/conf.py:
```diff
@@ -3,15 +3,13 @@
-
-globals().update(build_package_configs(
-    project_name="astro_metadata_translator",
-    version=astro_metadata_translator.version.__version__))
```
Is there a reason that you're not exporting the version anymore?

timj replied:
I am following the current documenteer guidelines. I don't know why version is no longer used.

fred3m commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -0,0 +1,248 @@
+                if "(104)" in str(e):
+                    raise FileNotFoundError(f"No such file or directory: {file}") from e
+                raise e
+            return None
+
+except ImportError:
+    from astropy.io import fits
+
+    def read_fits_metadata(file, hdu, can_raise=False):
+        """Read a FITS header using astropy.
+
+        Parameters
+        ----------
+        file : `str`
+            The file to read.
+        hdu : `int`
+            The header number to read.
+        can_raise : `bool`, optional
+            Indicate whether the function can raise and exception (default)
+            or should return `None` on error. Can still raise if an unexpected
+            error is encountered.
+
+        Returns
+        -------
+        md : `dict`
+            The requested header. `None` if it could not be read and
+            ``can_raise`` is `False`.
+        """
+        header = None
+        try:
+            with fits.open(file) as fits_file:
+                try:
+                    header = fits_file[hdu].header
+                except IndexError as e:
+                    if can_raise:
+                        raise e
+        except Exception as e:
+            if can_raise:
+                raise e
+        return header
```
I don't really like this block (mostly due to two functions having the exact same docstring) but I don't know that I have a _better_ suggestion. The best that I can come up with is to rename both of the above to `_read_fits_metadata` and outside of the try..except block have a
```
def read_fits_metadata(file, hdu, can_raise=False):
    """docstring here"""
    return _read_fits_metadata(file, hdu, can_raise)
```

but that seems overkill just to prevent repeating a docstring. The other obvious alternative would be to wrap this entire block in a function with the docstring, but I like that it sets `read_fits_metadata` on import as opposed to each function call.

timj replied:
That function has existed like that for ages but shows up as a diff because of the move from one file to another. I added docstrings when I moved and yes it's a bit annoying to repeat oneself. This is really an internal function anyhow. There is no expectation that this is used outside of the command line tooling or test code.

fred3m replied:
Ok, I'm just channeling an old review by John Swinbank where he discouraged me from doing this.

fred3m commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -0,0 +1,248 @@
+            The header number to read.
+        can_raise : `bool`, optional
+            Indicate whether the function can raise and exception (default)
+            or should return `None` on error. Can still raise if an unexpected
```
Maybe specify that this checks for a `fits` error code 104.

fred3m commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -0,0 +1,248 @@
+    """
+    if file.endswith(".yaml"):
+        md = read_test_file(file,)
+        if hdrnum != 0:
+            # YAML can't have HDUs
+            hdrnum = 0
```
Nothing else is done to `md` so we should just return the value here

```suggestion
        return md
```

timj replied:
It is used later to skip the merging step.

fred3m commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -0,0 +1,248 @@
+        the file.
+    """
+    if file.endswith(".yaml"):
+        md = read_test_file(file,)
```
Perhaps `md` -> `metadata`? (here and elsewhere in this module)

timj replied:
Sure, in the rest of the stack code `md` is a common variable so I was mainly following afw convention.

fred3m commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -0,0 +1,248 @@
+    file : `str`
+        Name of file to read. Can be FITS or YAML. YAML must be a simple
+        top-level dict.
+    hdrnum : `int`
+        Header number to read. Only relevant for FITS. If greater than 1
+        it will be merged with the primary header.
```
Since `hdrnum` is not required and ignored for YAML files I recommend giving it a default value of `0`.

timj replied:
it's not defaulted here but comes from the click command. I think I defaulted it to 1 because there were files around the place that didn't work by default with no merging.  The YAML option here is mostly to support internal testing and I don't expect users to be using YAML at all.

fred3m commented on python/astro_metadata_translator/bin/writeindex.py:
```diff
@@ -0,0 +1,108 @@
+        Regular expression string used to filter files when a directory is
+        scanned.
+    hdrnum : `int`
+        The HDU number to read. The primary header is always read and
```
and...?

fred3m commented on python/astro_metadata_translator/bin/writeindex.py:
```diff
@@ -0,0 +1,108 @@
+    if outpath is not None:
+        _, ext = os.path.splitext(outpath)
+        if ext != ".json":
+            raise ValueError(f"Override output file must end in .json but given {outpath}")
```
Neat, I didn't know that you can prepend a string with `f` instead of using `.format`!

timj replied:
Python 3.6 I think.

fred3m commented on python/astro_metadata_translator/bin/writeindex.py:
```diff
@@ -0,0 +1,108 @@
+
+    failed = []
+    okay = []
+    by_directory = {}
```
I'm not crazy about this variable name. `files_by_dir` might be more clear.

fred3m commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -0,0 +1,248 @@
+        ``obsInfo`` returns an ObsrvationInfo;
+        ``json`` returns a JSON string of the ObservationInfo;
+        ``metadata`` returns the metadata (will be unfixed form);
+        ``jsonmetadata`` returns the unfixed metadata in JSON string.
```
I wonder if this should be split into two parameters: `contentType`, which specifies either `obsInfo`, `obsSimple`, or `metadata`; and `format`, which specifies either `dict` or `json`. I think that there is a clear separation between the two and it will prevent you from having to do things like change the `mode` from `jsonmetadata`, etc. It also gives you the advantage that if you decided to add an additional format option in the future (yaml, pickle, etc) the implementation should be trivial.

fred3m commented on python/astro_metadata_translator/bin/writesidecar.py:
```diff
@@ -0,0 +1,131 @@
+from ..file_helpers import find_files, read_file_info
+
+
+def write_sidecar_json(file, hdrnum, mode, print_trace, outstream=sys.stdout, errstream=sys.stderr):
```
I would call this `write_sidecar_file` to show the connection and difference between this and `write_sidecar_files`.

fred3m commented on python/astro_metadata_translator/cli/astrometadata.py:
```diff
@@ -0,0 +1,233 @@
+@click.group(name="astrometadata", context_settings=dict(help_option_names=["-h", "--help"]))
+@click.option("--log-level",
+              type=click.Choice(["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"], case_sensitive=False),
+              default="INFO",
```
Isn't the default log level usually `WARNING`? Is there a particular reason that you're using a more verbose default?

timj replied:
I think for command line tooling most people prefer to see some feedback by default in the shell that their command is doing something.

fred3m replied:
fair enough

fred3m commented on python/astro_metadata_translator/cli/astrometadata.py:
```diff
@@ -0,0 +1,233 @@
+              help="When looking in a directory, regular expression to use to determine whether"
+              f" a file should be examined. Default: '{re_default}'")
+@click.pass_context
+def translate(ctx, files, quiet, hdrnum, mode, regex):
+
+    # For quiet mode we want to translate everything but report nothing.
+    if quiet:
+        mode = "none"
+
+    okay, failed = translate_header(files, regex, hdrnum, ctx.obj["TRACEBACK"], output_mode=mode)
+
+    if failed:
+        click.echo("Files with failed translations:", err=True)
+        for f in failed:
+            click.echo(f"\t{f}", err=True)
+
+    if not okay:
+        # Good status if anything was returned in okay
+        raise click.exceptions.Exit(1)
```
I'm curious about your decision to use `click` for the `astrometadata` interface while using `argparse` for `translateheader`, even though they both call the `process_files` function with the same set of arguments. Wouldn't it make more sense to remove argparse completely and just call this method from `translateheader.main`?

timj replied:
As I say elsewhere, click is the new code. The argparse in translate_header is the old implementation that I'm keeping around for backwards compatibility. I will add a deprecation message to it to let people know about the new click-based command and then will delete the argparse implementation.

fred3m replied:
That's fine if you just want to deprecate this command. But my thinking was that you could still support the old command line function by just pointing to the new click interface. Either way is fine with me.

timj replied:
I don't want to support the old translate_header.py command line. click and translate_header.py use the same implementation internally.

fred3m commented on python/astro_metadata_translator/cli/astrometadata.py:
```diff
@@ -0,0 +1,233 @@
+@click.option("-n", "--hdrnum", default=1,
+              help="HDU number to read. If the HDU can not be found, a warning is issued but translation"
+              " is attempted using the primary header. The primary header is always read and merged with"
+              " this header.")
```
There is a lot of repetition here, since most of the commands have the same options with the same default and help string. I wonder if it would be better to define them once and then just reuse, which would make updating any of them easier. For example
```
hdrnumArgs = ("-n" "--hdrnum")
hdrnumKwargs = {
    "default" : 1,
    "help": "HDU number to read..."
}
```

then later

```
@click.option(*hdrnumArgs, **hdunumKwargs)
```

for each sub-command that requires `hdrnum`.

fred3m commented on python/astro_metadata_translator/headers.py:
```diff
@@ -47,7 +48,14 @@ def merge_headers(headers, mode="overwrite", sort=False, first=None, last=None):
         - ``'overwrite'`` : Value in later header overwrites earlier value.
-        - ``'drop'`` : Entire key is dropped.
+        - ``'drop'`` : Entire key is dropped. If a key only appears in a
+          subset of the headers it will be retained.
```
By my read this isn't quite the correct behavior. Maybe something like "The entire key is dropped if it appears in more than one header and the value in all of the headers containing this key are not the same."

timj replied:
I have rephrased.

fred3m commented on python/astro_metadata_translator/headers.py:
```diff
@@ -47,7 +48,14 @@ def merge_headers(headers, mode="overwrite", sort=False, first=None, last=None):
+          ``__DIFF__``.  The order used matches the supplied order or
+          the sorted order if specified.  This allows a simple header diff
+          to be performed and associated with the original headers. Only
+          keys that appear in all headers will be retained in the merged one.
```
I would also change this entry to cover the case where a key is contained in a subset of headers, in which case it is also dropped and added to the `__DIFF__`.

fred3m commented on python/astro_metadata_translator/indexing.py:
```diff
@@ -0,0 +1,329 @@
+
+    Returns
+    -------
+    index_ : `ObservationGroup` or `dict[str, Union[dict, ObservaitonInfo]]`
+       If the index file referred to `ObservationInfo` this will return
+       an `ObservationGroup`, otherwise a `dict` will be returned with the
+       keys being paths to files and the values being the keys and values
+       stored in the index (with common information merged in). If
+       ``force_dict`` is `True` a `dict` will be returned with filename
+       keys even if the index file refers to `ObservationInfo` (the values
+       will be `ObservationInfo`).
```
I would point to the `process_index_data` docs instead of just copying it here.

fred3m reviewed (changes requested):
I left a few comments/suggestions below on the API and docstrings.

fred3m commented:
Oh, and I almost forgot to ask if there is any reason why `python/astro_metadata_translator/cli/__init__.py` is empty? How does it even import?

timj commented:
Couple of quick comments before I respond more specifically:

* Empty `__init__.py` files are fine. We don't need to export anything because the code is always using the explicit path.
* The old translate_header.py implementation is still there. I have not removed it yet because I firstly wanted to check that everything worked and secondly I was wondering if I should have a deprecation period where I make translate_header.py issue a warning. That is why the argparse code is still in that file.

timj commented:
@fred3m did you have an opinion on whether the `write-index` and `write-metadata-index` commands should be unified to a single `write-index` with a mode option? (similarly for sidecar writing).

fred3m commented:
I don't have a strong opinion on that since it depends a lot on the typical use case and I don't know what that is. If you feel like a substantial number of users (>10%?) will commonly want to write both together then I would make it easy on them and combine the commands. But if writing them together is a real outlier then I wouldn't worry about it.

timj commented:
You can't write both types of index or sidecar file together -- they have the same output file name. Maybe that is a compelling enough argument to combine them.

timj committed:
Add deprecation log message to old translate_header command

timj committed:
Factor out click options that are reused

timj committed:
Remove separate metadata subcommands

Now have a content parameter for write-sidecar and write-index
that can be "translated" (default) or "metadata".

Most of the changes here are to switch the parameter in the APIs
from mode to content and support "translated" as an option.

timj committed:
Change read_file_info to separate content from type of content in API

timj committed:
Make the fits metadata reader private and remove docstring duplication

timj committed:
Make sure that YAML reading is consistent with FITS reading

timj committed:
Allow translate command to raise on read and improve message

timj committed:
Fix docstring

timj committed:
Replace by_N variables with x_by_n

timj committed:
Fix test for change in translate_header error message

timj committed:
Rename write_sidecar_json to write_sidecar_file

timj committed:
Some docstring fixes to merge_headers

timj committed:
Update docstring for read_index

timj committed:
Change content to content_mode variable names

It was being confused with "content" meaning the actual content.

timj commented:
Thanks for taking another look. For others following along I've now added a `--content` argument to `write-sidecar` and `write-index` to switch between "translated" and "metadata" content.
